store/project_member.go

```diff
@@ -123,70 +123,121 @@ func (s *ProjectMemberService) DeleteProjectMember(ctx context.Context, delete *
+	if err != nil {
+		return nil, nil, err
+	}
 
 	oldMemberMap := make(map[int]*api.ProjectMember)
-	for _, existingMember := range existingProjectMemberList {
-		oldMemberMap[existingMember.PrincipalID] = existingMember
-	}
-
-	newMemberMap := make(map[int]bool)
-	for _, createMember := range set.List {
-		newMemberMap[createMember.PrincipalID] = true
-	}
-
-	createdMemberList := make([]*api.ProjectMember, 0)
-	deletedMemberList := make([]*api.ProjectMember, 0)
-	for _, memberCreate := range set.List {
-		// if the member exists, we will try to update its field
-		if memberBefore, ok := oldMemberMap[memberCreate.PrincipalID]; ok {
-			// if we update a member, we will the member in both createdMemberList and deletedMemberList
-			updatedMember, err := patchProjectMember(ctx, tx.PTx, &api.ProjectMemberPatch{
-				ID:           memberBefore.ID,
-				UpdaterID:    memberCreate.CreatorID,
-				Role:         (*string)(&memberCreate.Role),
-				RoleProvider: (*string)(&memberCreate.RoleProvider),
-				Payload:      &memberCreate.Payload,
-			})
-			if err != nil {
-				return nil, nil, FormatError(err)
-			}
-			// we append the updated member to createdMemberList, old member to the deletedMemberList
-			createdMemberList = append(createdMemberList, updatedMember)
-			deletedMemberList = append(deletedMemberList, memberBefore)
-		} else {
-			createdMember, err := createProjectMember(ctx, tx.PTx, memberCreate)
-			if err != nil {
-				return nil, nil, FormatError(err)
-			}
-			createdMemberList = append(createdMemberList, createdMember)
-		}
+	oldPrincipalIDList := make([]int, 0, len(oldProjectMemberList))
+	for _, oldMember := range oldProjectMemberList {
+		oldMemberMap[oldMember.PrincipalID] = oldMember
+		oldPrincipalIDList = append(oldPrincipalIDList, oldMember.PrincipalID)
 	}
-	for _, member := range existingProjectMemberList {
-		// if the member exists, we will try to update
-		if _, ok := newMemberMap[member.PrincipalID]; ok {
-			continue
+	newMemberMap := make(map[int]*api.ProjectMemberCreate)
+	newPrincipalIDList := make([]int, 0, len(batchUpdate.List))
+	for _, newMember := range batchUpdate.List {
+		newMemberMap[newMember.PrincipalID] = newMember
+		newPrincipalIDList = append(newPrincipalIDList, newMember.PrincipalID)
+	}
+
+	createPrincipalIDList, patchPrincipalIDList, deletePrincipalIDList, err := getBatchUpdatePrincipalIDList(ctx, oldPrincipalIDList, newPrincipalIDList)
+	if err != nil {
+		return nil, nil, FormatError(err)
+	}
+
+	createdMemberList = make([]*api.ProjectMember, 0)
+	deletedMemberList = make([]*api.ProjectMember, 0)
+	for _, id := range createPrincipalIDList {
+		memberCreate := newMemberMap[id]
+		createdMember, err := createProjectMember(ctx, tx.PTx, memberCreate)
+		if err != nil {
+			return nil, nil, FormatError(err)
+		}
+		createdMemberList = append(createdMemberList, createdMember)
+	}
+
+	for _, id := range patchPrincipalIDList {
```

Comment:
With the new logic, if we are patching (e.g. re-sync the GitLab and noticed the role change), then we should create an updated role activity instead of 1 created, 1 deleted activity.

Reply:
the logic in server/project_member.go would take care of this